compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/SplitReduction.cpp

```diff
@@ -389,4 +390,352 @@ splitReduction(RewriterBase &rewriter, LinalgExt::TopkOp topkOp,
+
+  SmallVector<utils::IteratorType> newIteratorTypes =
+      genericOp.getIteratorTypesArray();
+  newIteratorTypes.insert(newIteratorTypes.begin() + insertSplitDimension,
+                          utils::IteratorType::parallel);
+
+  SmallVector<AffineMap> newMaps;
+  unsigned numInputs = genericOp.getNumDpsInputs();
+  for (OpOperand &operand : genericOp->getOpOperands()) {
+    AffineMap map = genericOp.getMatchingIndexingMap(&operand);
+    if (operand.getOperandNumber() < numInputs) {
+      newMaps.push_back(getExpandedInputIndexingMap(
+          map, reductionDim, insertSplitDimension, control.innerParallel,
+          rewriter.getContext()));
+    } else {
+      newMaps.push_back(getExpandedOutputIndexingMap(map, insertSplitDimension,
+                                                     rewriter.getContext()));
+    }
+  }
+
+  // Create partial linalg.generic op with global index computation.
+  Value tileSize = rewriter.create<arith::ConstantIndexOp>(loc, ratio);
+  auto partialOp = rewriter.create<linalg::GenericOp>(
+      loc, TypeRange{identityValue.getType(), identityIndex.getType()},
+      newInputs, ValueRange{identityValue, identityIndex}, newMaps,
+      newIteratorTypes);
+
+  rewriter.inlineRegionBefore(genericOp.getRegion(), partialOp.getRegion(),
+                              partialOp.getRegion().begin());
+
+  Block &body = partialOp.getRegion().front();
+  rewriter.setInsertionPointToStart(&body);
+
+  unsigned innerIdxDim = reductionDim + 1;
+  unsigned outerIdxDim = insertSplitDimension;
+
+  // Compute global index (gidx) for reduction when the original reduction
+  // dimension is split into [outerIdx, innerIdx] using `ratio`. This is used to
+  // correctly compute the global index for comparisons and index selection.
+  Value outerIdx = rewriter.create<linalg::IndexOp>(loc, outerIdxDim);
+  Value innerIdx = rewriter.create<linalg::IndexOp>(loc, innerIdxDim);
+  Value offset = rewriter.create<arith::MulIOp>(loc, outerIdx, tileSize);
+  Value gidx = rewriter.create<arith::AddIOp>(loc, offset, innerIdx);
+
+  auto selectOp = dyn_cast<arith::SelectOp>(combinerOps.selectOp);
+  Value oldIdx = selectOp.getTrueValue();
+  Value newIdx = gidx;
+  if (oldIdx.getType() != gidx.getType()) {
+    newIdx = rewriter.create<arith::IndexCastOp>(loc, oldIdx.getType(), gidx);
+  }
+  selectOp.setOperand(1, newIdx);
+  rewriter.setInsertionPointAfter(partialOp);
+
+  unsigned intermRank = expandedValueShape.size();
+  AffineMap valueMap = rewriter.getMultiDimIdentityMap(intermRank);
+  AffineMap indexMap = valueMap;
+
+  SmallVector<AffineExpr> resultExprs;
+  SmallVector<utils::IteratorType> reductionIteratorTypes(
+      intermRank, utils::IteratorType::parallel);
+  reductionIteratorTypes[insertSplitIndex] = utils::IteratorType::reduction;
+  for (unsigned i : llvm::seq<unsigned>(0, intermRank)) {
+    if (i != insertSplitIndex)
+      resultExprs.push_back(rewriter.getAffineDimExpr(i));
+  }
```

Comment:
style nit: use auto because `llvm::seq<>` already spells the type.
style nit 2: wrap simple single statement with braces.

```suggestion
  for (auto i : llvm::seq<unsigned>(0, intermRank)) {
    if (i != insertSplitIndex) {
      resultExprs.push_back(rewriter.getAffineDimExpr(i));
    }
  }
```